File: Qual/2_22_1/21_22_1_QE_prob1_old.cpp
#include <iostream>
#include <string>
#include <vector>
#include <algorithm>

bool palindrome(const std::string& s) {
    return s == std::string(s.rbegin(), s.rend());
}

bool substring(const std::string& s, const std::string& t) {
    return s.find(t) != std::string::npos;
}

std::vector<std::string> max_palindromes(const std::string& s) {
    int n = s.length();
    int w = n;
    std::vector<std::string> pallist;

    while (w) {
        for (int i = 0; i < n - w + 1; ++i) {
            int flag = 0;
            if (palindrome(s.substr(i, w))) {
                flag = 1;
                for (const std::string& p : pallist) {
                    if (substring(p, s.substr(i, w))) {
                        flag = 0;
                        break;
                    }
                }
            }
            if (flag == 1) {
                pallist.push_back(s.substr(i, w));
            }
        }
        w -= 1;
    }

    // Sorting the output
    std::vector<std::string> out;
    for (int i = 0; i < n; ++i) {
        for (int j = i + 1; j <= n; ++j) {
            std::string sub = s.substr(i, j - i);
            if (std::find(pallist.begin(), pallist.end(), sub) != pallist.end()) {
                out.push_back(sub);
            }
        }
    }

    return out;
}

int main() {
    std::string s1 = "kabccbadzdefgfeda";
    std::string s2 = "kabccba dzabccbaza";
    std::string s3 = "asdsa azazza";

    auto print_result = [](const std::string& s) {
        std::vector<std::string> result = max_palindromes(s);
        for (const auto& str : result) std::cout << str << " ";
        std::cout << std::endl;
    };

    print_result(s1);
    print_result(s2);
    print_result(s3);

    return 0;
}

File: Qual/2_22_1/21_22_1_QE_prob1_old.py
def palindrome(s: str)->bool:
    return s == s[::-1]

def substring(s:str, t:str)->bool:
    return t in s

def max_palindromes(s:str)->list:
    n = len(s); w = len(s)
    pallist= []
    while w:
        for i in range(n-w+1):
            flag = 0
            if palindrome(s[i:i+w]):
                flag = 1
                for j in range(len(pallist)):
                    if(substring(pallist[j], s[i:i+w])):
                        flag = 0
            if (flag == 1):
                pallist.append(s[i:i+w])
        w -= 1

    # 순서 맞추기
    out = []
    for i in range(n):
        for j in range(i+1, n+1):
            if s[i:j] in pallist:
                out.append(s[i:j])

    return out

s1 = "kabccbadzdefgfeda"
s2 =  "kabccba dzabccbaza"
s3 = "asdsa azazza"
print(max_palindromes(s1))
print(max_palindromes(s2))
print(max_palindromes(s3))